### PR TITLE
Add PostgresqlCheck actor

### PR DESCRIPTION
**Pull request by panovotn**
This actor collects information about Red Hat Signed PostgreSQL
installation and warns user about possible manual intervention
requirements where applicable.

### CODE REVIEW

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (hhorak) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: We should ideally link to internal upgrade steps (currently available only in 8.0 Beta docs), and we don't know a correct link that will be ready after GA. How to handle this? Any idea, @pirat89 @bocekm?

Comment (vinzenz) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: @hhorak just post a patch once you have it, or have an RedHat KB article you can with some text and update the link in that article once you have the right link

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (hhorak) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: This seems to be an issue, the superuser is called `postgres`. Is it working this way?

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (hhorak) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: I'd be be more concrete here. For example:
"PostgreSQL server component will be upgraded. Since RHEL-8 includes PostgreSQL server 10 by default, which is incompatible with 9.2 included in RHEL-7, it is necessary to proceed with additional steps for complete upgrade of the PostgreSQL data. Back-up your data before proceeding with the upgrade and read more about upgrade steps at <KB_article_url>."

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (hhorak) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Wouldn't False here cause error later, when the output of this function is taken as array? Maybe returning an empty array in this case would be better.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (vinzenz) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]:
```suggestion
            if extensions_removed:
```

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (vinzenz) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]:
```suggestion
        if pkg.name == 'postgresql-contrib' and extensions:
```

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (vinzenz) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Its probably even better to return None as it’s not an empty list of extensions but the command failed

If you check if value instead of if len(value) then this will work with false none and an empty list

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Definitely isn't. This happened while renaming get_postgresql_packages function. Thanks for pointing it out. Fixed.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Fixed. Thanks.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Fixed as well.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: I agree. Returning None seems much more universal as @vinzenz pointed out. Fixed.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: I've used text you've provided here (thank you), and included a KB article URL. That article is empty ATM, but will be filled in as soon as I get proper permissions. This should be sufficient for the sake of merging this PR.

**Comment by UncleAlbie** [comment]
As I've addressed all comments, can we merge?

**Comment by pirat89** [comment]
@UncleAlbie If I see right, there are still unresolved discussions - between those,at least one has not been adressed.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: As I've mentioned in another conversation, the URL is now changed to the KB article. This should be fixed as well.

**Comment by UncleAlbie** [comment]
@pirat89 Those have been fixed as well, I just mentioned those fixes only in one of the conversations. My mistake. Now it should be ready.

**Comment by hhorak** [comment]
@pirat89 just checking whether there is something else that needs to be done before merging. @UncleAlbie claims the requested changes were addressed.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (vinzenz) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: This seems to be an error:
```suggestion
            '-c', 'psql -tAc \'\\dx\''
```

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (vinzenz) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]:
```suggestion
            extensions_removed = [ext for ext in extensions if ext['name'] in MISSING_EXTENSIONS]
```

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (vinzenz) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]:
```suggestion
    extensions_public = [ext for ext in extensions if ext['schema'] == 'public']
```

Comment (vinzenz) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: And out of curiosity, who says that the extensions are installed to public schema? Wouldn't it be possible to have it installed to a different schema? I mean I know it's common to do it in the public schema but I am not sure if someone might use a different schema than the public schema, since the CREATE EXTENSION has a schema parameter.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (vinzenz) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Are you sure that should be a debug message?

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Suggested change pushed as fixup.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Suggested change pushed as fixup.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Suggested change pushed as fixup.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Suggested change pushed as fixup.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Suggested change pushed as fixup.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Suggested change pushed as fixup.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Suggested change pushed as fixup.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Changed to 'warning' and pushed as fixup.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: You're right. However, I'd suggest solving this in another PR as the public schema really is commonly used.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (vinzenz) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Needs check for ‘extensions’ be Ayse it can be None

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (shaded-enmity) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Please use `report_with_links` if you want to include links - https://github.com/oamg/leapp-repository/blob/master/repos/system_upgrade/el7toel8/libraries/reporting.py#L40

**Comment by pirat89** [comment]
@UncleAlbie Please, do not forget to rebase against the current master when you resolve the additional required changes.

**Review by vinzenz — DISMISSED** [review]
Please apply the comments from Pavel and me and once more rebase against current master. Thanks

**Comment by vinzenz** [comment]
Ping? Please address comments and rebase, otherwise this PR is almost ready.

**Comment by UncleAlbie** [comment]
I just need to finalize tests for the actor. Hopefully, I'll get to it by the end of this week.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Fixed. Thanks for pointing that out.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: I've changed the report to `report_with_links`.

**Comment by hhorak** [comment]
There is also the article that needs to be published before this PR is merged: https://access.redhat.com/articles/4055641

**Comment by UncleAlbie** [comment]
Tests for this actor is now failing on "reporting" module import .. could this be a LEAPP issue, or am I missing something?

**Comment by bocekm** [comment]
We have changed the reporting feature in the leapp framework. Please take a look at this PR on examples how to update reporting function calls:
https://github.com/oamg/leapp-repository/pull/295/files
Here's a documentation on how to use the new reporting:
https://leapp.readthedocs.io/en/latest/actor-rhel7-to-rhel8.html#producing-data-for-other-actors-and-reporting

**Comment by bocekm** [comment]
leapp-ci build

**Comment by bocekm** [comment]
@fellipeh, @asilveir, you can take a look at this PR if you have time. Thanks.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (pirat89) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Currently, only RPMs that are checked in the actor are `postgresql-server` and `postgresql-contrib`. From that point it doesn't seem necessary to use this function. Instead of that, wouldn't be better to use `has_package` function, that will provide that info? See:

```
   from leapp.libraries.common.rpms import has_package

   if has_package(InstalledRedHatSignedRPM, 'postgresql-server'):
```

We prefer this library function to be used instead, but it is not blocker. The code looks good anyway, but I haven't tested it..

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/tests/test_postgresqlcheck.py
Comment (pirat89) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/tests/test_postgresqlcheck.py]: The `produce_mocked` function is not ideal to test produced reports and it is duplicit code. Similar to `create_report_mocked`, which is not completely same, but pretty similar. Look at `create_report_mocked` in `libraries/testutils.py` instead. There are already some actors using it. And you can remove the `produce_mocked` completely as `produce` is part of the `create_report ` function.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/tests/test_postgresqlcheck.py
Comment (pirat89) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/tests/test_postgresqlcheck.py]:
```suggestion
```
This is not needed here.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/tests/test_postgresqlcheck.py
Comment (pirat89) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/tests/test_postgresqlcheck.py]:
```suggestion
    monkeypatch.setattr(api, "produce", produce_mocked())
```

Same here.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/tests/test_postgresqlcheck.py
Comment (pirat89) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/tests/test_postgresqlcheck.py]: I prefer to create the `packages` input directly (if it is valid after possible changes in the actor I mentioned earlier). It doesn't make so much sense to test it in this way. If you want to keep the get_postgresql_packages function, test it separately please.

**Comment by UncleAlbie** [comment]
I've rewritten the actor from the ground up. There's been some changes in the framework and it just made sense to me. It's much simpler now.

As all checks have passed, it should be ready for review. However, please, don't merge yet. There's still some work around the KB article needed. Thanks!

**Comment by pirat89** [comment]
@UncleAlbie thanks for info. So we will do review without merging just to give you feedback.

**Comment by shaded-enmity** [comment]
@oamg/developers @UncleAlbie Are there any blockers? The code looks very good to me.

**Comment by bocekm** [comment]
@UncleAlbie, is this PR still in progress? Is it ready for another round of reviews?

**Comment by bocekm** [comment]
@UncleAlbie, ping :)

**Comment by Rezney** [comment]
@UncleAlbie is there anything we can do to move this forward? Thanks...

**Comment by UncleAlbie** [comment]
This PR should be ready for (review and) merge.

**Review by pirat89 — CHANGES_REQUESTED** [review]
@UncleAlbie I apologize for really long waiting for review. I've just discovered couple of things now:
- the actors library needs to be renamed to ensure the name is uniq - it's to ensure the unit-tests can be executed correctly in faster way, instructions about naming of test files and libraries are here:
- https://leapp.readthedocs.io/en/latest/unit-testing.html#naming-conventions
- https://leapp.readthedocs.io/en/latest/unit-testing.html#testing-private-actor-library-unit-tests

In short, call actor's library e.g. "postgresqlcheck.py" and fix imports. 

Other stuff in comments/suggestions below. I think that the rest seems fine after the stuff will be fixed. I will give it another look tomorrow during proper work-time.

**Comment by pirat89** [comment]
Note: the article is not published yet.

**Comment by Rezney** [comment]
@pirat89 is there anything we can do to move this forward?

**Comment by pirat89** [comment]
@Rezney  the referred article is not published yet (https://access.redhat.com/articles/4055641), probably some additional refactoring is needed and several required changes have not been addressed. We can sync with guys to push it, but currently there is not so much time. Probably we could sync with @hhorak (ping, wdyt Honzo?) about this PR to decide what to do with it (finish it? close it?). In case of the finalisation work, we should schedule a timeplan to ensure that someone from our team and someone from database team have time to finish this. As well we should sync about the testing.

**Comment by hhorak** [comment]
We agreed with @pirat89 and @panovotn to finish this in 1st part of June, the article should be published till then.
Also tracking this change in BZ https://bugzilla.redhat.com/show_bug.cgi?id=1962081 in order to make syncing with QE easier.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (panovotn) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Fixed according to your suggestion. Thanks.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (panovotn) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Fixed according to your suggestion. Thanks.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (panovotn) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Fixed according to your suggestion. Thanks.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (panovotn) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Fixed according to your suggestion. Thanks.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py
Comment (panovotn) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: The article content is actually duplicate to the upgrade section in our PostgreSQL documentation. Therefore, I've changed the URL to a link to that section.

IMO, separate article isn't needed for this version of the actor.

**Review by pirat89 — APPROVED** [review]
Lgtm. Regarding tests it seems it ok. I will test it manually yet.

**Comment by pirat89** [comment]
leapp-ci build

**Comment by pirat89** [comment]
leapp-ci build

**Comment by pirat89** [comment]
:snail: :turtle:  After two years and two months: :fireworks: :confetti_ball:  :partying_face:  :champagne:

**Comment by pirat89** [comment]
Btw, just for the completness, here is part of the report when postgresql-server and postgresql-contrib packages are installed:
```
----------------------------------------
Risk Factor: medium
Title: PostgreSQL (postgresql-server) has been detected on your system
Summary: PostgreSQL server component will be upgraded. Since RHEL-8 includes PostgreSQL server 10 by default, which is incompatible with 9.2 included in RHEL-7, it is necessary to proceed with additional steps for the complete upgrade of the PostgreSQL data.
Remediation: [hint] Back up your data before proceeding with the upgrade and follow steps in the documentation section "Migrating to a RHEL 8 version of PostgreSQL" after the upgrade.
Key: 6966486bbbf49710d1d9f4e4f6b612217aca38a1
----------------------------------------
Risk Factor: medium
Title: PostgreSQL (postgresql-contrib) has been detected on your system
Summary: Please note that some extensions have been dropped from the postgresql-contrib package and might not be available after the upgrade:
    - dummy_seclabel
    - test_parser
    - tsearch2
Key: e0703e8e14b8c2eee885de3fb451c9d072b89ed6
```

I've tested even the case when just the postgresql-server package is installed, so just the one report is generated as expected.

**Pull request merged**
